### PR TITLE
Extend functionality of `no-for-loop`

### LINKED ISSUE #286: `no-for-loop` false negative
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/286)

```js
/* eslint unicorn/no-for-loop: 2 */
for (let i = 0; i < arr.length; i += 1) {
  console.log(arr[i])
}
```

**Expected**:
It reports.
**Actual**:
It does not report.

This is a subtask of https://github.com/sindresorhus/eslint-plugin-unicorn/issues/250 .

> There is a $60.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/286).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by karaggeorge**
Fixes #286 

Works on part of #250 

## Summary
- Will track for variable declaration to `arrayName[indexName]` anywhere in the body
- Will fix even if the declaration is complex : `let x = 1, y = arr[i], z = 13;`
- Replaces all instances of `arr[i]` with the element name
- Will not report if `arr[i]` is set to anything or if the array variable is used any other way than to read `arr[i]` since that might have different effects in a regular `for` loop versus a `for..of`

- Moved the case of #252 to be completely ignored instead of not fixed since this is valid example of that, that can't be moved to `for..of`: 
```
for (let i = 0;i < arr.length;i++) {
  let el = arr[i];
  if (el.skipNext) {
    i++;
  }
  console.log(el);
}
```

So now, any time the index is set to in the loop body, the rule will ignore it

## Question

When there is this case: 
```
/* eslint unicorn/no-for-loop: 2 */
for (let i = 0; i < arr.length; i += 1) {
  console.log(arr[i])
}
```

it can be fixed with 
```
/* eslint unicorn/no-for-loop: 2 */
for (const value of arr) {
  console.log(value)
}
```

But the name `value` is arbitrary since the code doesn't provide a name itself. Do we want to keep it `value`? Or not fix? Or provide an option?

### CODE REVIEW

**Comment by sindresorhus** [comment]
> But the name value is arbitrary since the code doesn't provide a name itself. Do we want to keep it value? Or not fix? Or provide an option?

Lets call it `element`. It should be fixable. I don't see the need to customize that as the naming depends on the context. User's should just give it a proper name afterwards.

File: rules/no-for-loop.js [hunk 1]
```diff
@@ -304,9 +323,15 @@ const create = context => {
 							node.init.range[0],
 							node.update.range[1]
 						], replacement),
-
-						fixer.removeRange(getRemovalRange(elementVariableDeclaration, sourceCode))
-					];
+						...arrayReferences.map(reference => {
+							if (reference === elementReference) {
+								return null;
```
Comment (sindresorhus) [inline rules/no-for-loop.js]:
```suggestion
								return;
```

File: rules/no-for-loop.js [hunk 2]
```diff
@@ -304,9 +323,15 @@ const create = context => {
...
+							}
+
+							return fixer.replaceText(reference.identifier.parent, element);
+						}),
+						elementNode ? fixer.removeRange(getRemovalRange(elementNode, sourceCode)) : null
```
Comment (sindresorhus) [inline rules/no-for-loop.js]: I prefer using `undefined`.

```suggestion
						elementNode && fixer.removeRange(getRemovalRange(elementNode, sourceCode))
```

File: rules/no-for-loop.js (hunk 1 above)
Comment (karaggeorge) [reply rules/no-for-loop.js]: This is caught by XO:

`329:9   Arrow function expected a return value.  array-callback-return`

Comment (sindresorhus) [reply rules/no-for-loop.js]: Right, you can `return undefined;` then.

Comment (karaggeorge) [reply rules/no-for-loop.js]: 750aa1c

**Comment by sindresorhus** [comment]
@golopot Hehe. You were seconds too late.

@karaggeorge Can you handle that?

**Comment by karaggeorge** [comment]
@sindresorhus @golopot Yeah, that's weird. I'll fix and open a new PR for it 😄

**Comment by karaggeorge** [comment]
@sindresorhus @golopot Ended up being an issue of getting the references in a scope (doesn't include child scopes). Fixed here: https://github.com/sindresorhus/eslint-plugin-unicorn/pull/303

**Pull request merged**